libs/model/src/thread/UpdateThread.command.ts

```diff
@@ -0,0 +1,377 @@
+  actor: Actor,
+  auth: AuthContext,
+  {
+    locked,
+    archived,
+    stage,
+    topic_id,
+  }: z.infer<typeof schemas.UpdateThread.input>,
+) {
+  const patch: Partial<ThreadAttributes> = {};
+
+  if (typeof locked !== 'undefined') {
+    patch.read_only = locked;
+    patch.locked_at = locked ? new Date() : null;
+  }
+
+  typeof archived !== 'undefined' &&
+    (patch.archived_at = archived ? new Date() : null);
+
+  if (typeof stage !== 'undefined') {
+    const community = await models.Community.findByPk(auth.community_id!);
+    mustExist('Community', community);
+
+    const custom_stages =
+      community.custom_stages.length > 0
+        ? community.custom_stages
+        : ['discussion', 'proposal_in_review', 'voting', 'passed', 'failed'];
+
+    if (!custom_stages.includes(stage))
+      throw new InvalidInput(UpdateThreadErrors.InvalidStage);
+
+    patch.stage = stage;
+  }
+
+  if (typeof topic_id !== 'undefined') {
+    const topic = await models.Topic.findOne({
+      where: { id: topic_id, community_id: auth.community_id! },
+    });
+    mustExist('Topic', topic);
+
+    patch.topic_id = topic_id;
+  }
+
+  if (Object.keys(patch).length > 0) {
+    const authorized =
+      actor.user.isAdmin ||
+      ['admin', 'moderator'].includes(auth.address!.role) ||
+      auth.is_author;
+    if (!authorized)
+      throw new InvalidActor(actor, 'Must be admin, moderator, or author');
+  }
+  return patch;
+}
+
+export function UpdateThread(): Command<
+  typeof schemas.UpdateThread,
+  AuthContext
+> {
+  return {
+    ...schemas.UpdateThread,
+    auth: [isAuthorized({ collaborators: true })],
+    body: async ({ actor, payload, auth }) => {
+      const { address } = mustBeAuthorized(actor, auth);
+      const { thread_id, discord_meta } = payload;
+
+      // find by discord_meta first if present
+      const thread = await models.Thread.findOne({
+        where: discord_meta ? { discord_meta } : { id: thread_id },
+      });
+      if (!thread) throw new InvalidInput(UpdateThreadErrors.ThreadNotFound);
+
+      const content = getContentPatch(thread, payload);
+      const adminPatch = getAdminOrModeratorPatch(actor, auth!, payload);
+      const ownerPatch = await getAdminOrModeratorOrOwnerPatch(
+        actor,
+        auth!,
+        payload,
+      );
+      const collaboratorsPatch = await getCollaboratorsPatch(
+        actor,
+        auth!,
+        payload,
+      );
+
+      // check if patch violates contest locks
+      if (
+        Object.keys(content).length > 0 ||
+        ownerPatch.topic_id ||
+        collaboratorsPatch.add.length > 0 ||
+        collaboratorsPatch.remove.length > 0
+      ) {
+        const contestManagers = await models.sequelize.query(
+          `
+SELECT cm.contest_address FROM "Threads" t
+JOIN "ContestTopics" ct on ct.topic_id = t.topic_id
+JOIN "ContestManagers" cm on cm.contest_address = ct.contest_address
+WHERE t.id = :thread_id
+`,
+          {
```

Comment:
Query looks fine. There's a helper called `getThreadContestManagers` but it only gets the *active* contest managers, while in this context, we want to check if the thread was ever part of a contest.